Configure logging to log to Stackdriver (when flag is set)

carise opened the pull request:
Logger has a console handler (logging.StreamHandler) and a syslog handler (logging.handlers.SysLogHandler). On GCE instance we install the fluentd logging agent which will transparently pass the syslog entries to stackdriver

resolves #73

ahoying commented on google/cloud/security/common/gcp_api/compute.py:
```diff
@@ -14,11 +14,37 @@
+
 from google.cloud.security.common.gcp_api._base_client import _BaseClient
 
+def is_compute_engine_instance():
```
It seems a bit out of place to put this here. I think a module under ../util would be better.

ahoying commented on google/cloud/security/common/third_party/transports/background_thread.py:
```diff
@@ -0,0 +1,171 @@
+# Copyright 2016 Google Inc.
```
How will this stay in sync with the file in google-cloud-python?

ahoying commented on google/cloud/security/common/util/log_util.py:
```diff
@@ -12,29 +12,90 @@
+FLAGS = flags.FLAGS
+flags.DEFINE_boolean('use_cloud_logging', False,
+                     'Use Cloud Logging, if available.')
+flags.DEFINE_boolean('nouse_cloud_logging', False, 'Do not use Cloud Logging.')
```
DEFINE_boolean automatically creates the no<blah> version of the flag, you don't have to do that yourself. The no<blah> version will set the flag value False.

And based on the code below, the default value should be None, not False.

ahoying commented on google/cloud/security/common/util/log_util.py:
```diff
@@ -12,29 +12,90 @@
+
+    # Determine whether to use cloud logger based on environment.
+    if should_use_cloud_logger is None:
+        should_use_cloud_logger = is_gce
```
Can this be simplified to not require the should_use_cloud_logger var and just the following.

if FLAGS.use_cloud_logging or (FLAGS.use_cloud_logging is None and is_gce):
...

ahoying commented on google/cloud/security/common/util/log_util.py:
```diff
@@ -12,29 +12,90 @@
+        An instance of the configured logger.
+    """
+    formatter = logging.Formatter(LOG_FORMAT)
+    console_handler = logging.StreamHandler()
```
Should we default to SysLogHandler instead? Or maybe use logging.config.fileConfig and include a default logging configuration that can be overridden by changing the configuration file rather than changing the code?

mirons-google commented on google/cloud/security/common/data_access/dao.py:
```diff
@@ -29,11 +29,8 @@
 from google.cloud.security.common.data_access.errors import NoResultsError
 from google.cloud.security.common.data_access.sql_queries import create_tables
 from google.cloud.security.common.data_access.sql_queries import select_data
-from google.cloud.security.common.util.log_util import LogUtil
```
We don't need to add back in logging support?

carise replied:
it wasn't used anywhere in this file.

mirons-google commented on google/cloud/security/common/data_access/csv_writer.py:
```diff
@@ -18,11 +18,8 @@
 import tempfile
 
 from google.cloud.security.common.data_access.errors import CSVFileError
-from google.cloud.security.common.util.log_util import LogUtil
```
We don't need to add back in logging support?

ahoying commented on google/cloud/security/common/gcp_api/cloud_logging.py:
```diff
@@ -0,0 +1,443 @@
+import logging
+
+from google.cloud.security.common.gcp_api._base_client import _BaseClient
+# pylint: disable=line-too-long
```
You need to re-enable this check after the imports to catch the other places in this file where line is too long.

ahoying commented on google/cloud/security/common/gcp_api/cloud_logging.py:
```diff
@@ -0,0 +1,443 @@
+    # pylint: disable=line-too-long
+    """Cloud Logging Handler.
+
+    See: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/logging/google/cloud/logging/handlers/handlers.py
```
Is there an issue with just directly using google-cloud-python instead of importing pieces of it? How do we get patches / updates if there are bug fixes?

blueandgold replied:
+1 on not importing pieces of code like this if we can avoid it.  But the issue is different versions of oauth2client being used.  Have other people done something similar to here?

mirons-google commented on google/cloud/security/common/third_party/transports/base.py:
```diff
@@ -0,0 +1,35 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module containing base class for logging transport."""
```
Why don't we just pip install google-cloud-logging and use it as a dependency?

carise replied:
googleapiclient and google-cloud-python don't play together at all. installing both breaks the entire environment.

mirons-google commented on google/cloud/security/common/util/log_util.py:
```diff
@@ -12,29 +12,90 @@
+    console_handler.setFormatter(formatter)
+    logger_instance = logging.getLogger(module_name)
+    logger_instance.addHandler(console_handler)
+    logger_instance.setLevel(logging.INFO)
```
We might want to look into making this an argument at execution.

mirons-google commented on google/cloud/security/common/util/log_util.py:
```diff
@@ -12,29 +12,90 @@
+    Args:
+        module_name: The name of the module for which to configure logging.
+    """
+    (is_gce, _) = compute.is_compute_engine_instance()
```
See my comment above about not needing the error message. I suggest just returning the boolean result.

mirons-google commented on google/cloud/security/common/util/log_util.py:
```diff
@@ -12,29 +12,90 @@
+        logger_instance = logging.getLogger(module_name)
+        logger_instance.addHandler(cloud_handler)
+
+        if FLAGS.debug:
```
Why isn't this done in get_logger? It could be confusing to debug the code to understand where this is getting set if we set verbosity in two different spots.

carise replied:
you can't use flags until they are parsed, which happens in app.run(). So the strategy is to attach the default log handler at the toplevel (e.g. module import) and then attach the stackdriver handler after app.run().

carise commented:
I haven't cleaned up all the imports yet, mostly the ones that were nearby the log_util import.

Changed logger to use the SysLogHandler and removed the cloud logging related stuff. PTAL.

carise commented on tests/inventory/pipelines/base_pipeline_test.py:
```diff
@@ -87,11 +88,11 @@ def test_get_loaded_count(self):
 
-        self.pipeline.logger = mock.create_autospec(
-            LogUtil).setup_logging('foo')
+        base_pipeline.LOGGER = mock.create_autospec(
```
@blueandgold FYI I tweaked your pipeline unit tests to reference the [MODULE_NAME].LOGGER property and made the pipeline.logger property a global "LOGGER" variable to be consistent across the codebase. Let me know if this looks ok; I'm not totally sure what's the best way here.

blueandgold replied:
No problem if that works.  :+1:

blueandgold commented:
Adding `cla: yes` label, as the PR author is a googler at the time of this PR was created.
